README.md
# MailChimp3

[![CircleCI](https://circleci.com/gh/planningcenter/mailchimp3/tree/master.svg?style=svg)](https://circleci.com/gh/planningcenter/mailchimp3/tree/master)

`mailchimp3` is a Rubygem that provides a very thin, simple wrapper around the MailChimp RESTful JSON "Marketing" API version 3
documented at [mailchimp.com/developer/marketing/api](https://mailchimp.com/developer/marketing/api/).

This wrapper is very low-level -- you'll still be dealing with individual GET, POST, PATCH, and DELETE requests, but the gem
handles the OAuth2 flow (getting a token), passing the auth token in every request, and raises descriptive errors you can easily rescue
in your own code.

## Installation

```
gem install mailchimp3
```

## Usage with HTTP Basic Auth

1. Set your HTTP Basic auth key somewhere in your app
   (probably an initializer if using Rails):

   ```ruby
   MailChimp3.config.basic_auth_key = 'key-us2'
   ```

2. Create a new API object and use it:

   ```ruby
   # HTTP Basic
   api = MailChimp3.new
   ```

3. Call a method on the api object to build the endpoint path.

   ```ruby
   api.lists
   # /lists
   ```

4. For IDs, treat the object like a hash (use square brackets).

   ```ruby
   api.lists['abc123'].members
   # /lists/abc123/members
   ```

5. To execute the request, use `get`, `post`, `patch`, or `delete`, optionally passing arguments.

   ```ruby
   api.lists.get(count: 25)
   # GET /lists?count=25
   api.lists['abc123'].members['cde345'].get
   # GET /lists/abc123/members/cde345
   ```

### Usage with OAuth 2

1. Set your OAuth client id and secret somewhere in your app
   (probably an initializer if using Rails):

   ```ruby
   MailChimp3.config.client_id = 'abc123'
   MailChimp3.config.client_secret = 'xyz456'
   ```

2. (First time for each user) Get an OAuth 2 token by calling `MailChimp3.oauth.authorize_url` and redirect your user to it:

   ```ruby
   url = MailChimp3.oauth.authorize_url(
     redirect_uri: 'http://example.com/oauth/callback'
   )
   redirect_to url
   ```

3. Upon redirect back to your app (in your `/oauth/callback` action),
   call `MailChimp3.oauth.complete_auth`, passing in the `code` param
   and the `redirect_uri` again.

   ```ruby
   data = MailChimp3.oauth.complete_auth(
     params[:code],
     redirect_uri: 'http://example.com/oauth/callback'
   )
   ```

   The hash returned looks like this:

   ```ruby
   {
     token: <OAuth2::AccessToken>
     token_string: 'abc123',
     metadata: {
       dc: 'us2'
     }
   }
   ```

   Then get the authentication token and data center, and store it on
   your user record for later use:

   ```ruby
   user.update_attributes(
     mailchimp_auth_token: data[:token_string],
     mailchimp_data_center: data[:metadata][:dc]
   )
   ```

4. (Subsequent times for user) Instantiate the api object, passing in the
   auth token and data center:

   ```ruby
   api = MailChimp3.new(
     oauth_access_token: user.mailchimp_auth_token,
     dc: user.mailchimp_data_center
   )
   ```

5. Use the `api` instance to make API calls!

## Example

```ruby
require 'mailchimp3'

api = MailChimp3.new(basic_auth_key: 'abc123abc123abc123abc123abc123ab-us2')
api.lists.post(
  name: 'Church.IO',
  email_type_option: false,
  permission_reminder: 'signed up on https://church.io'
  contact: {
    company: 'TJRM',
    address1: '123 N. Main',
    city: 'Tulsa',
    state: 'OK',
    zip: '74137',
    country: 'US'
  },
  campaign_defaults: {
    from_name: 'Tim Morgan',
    from_email: 'tim@timmorgan.org',
    subject: 'Hello World',
    language: 'English'
  },
)
```

...which returns something like:

```ruby
{
  "id"   => "abc123abc1",
  "name" => "Church.IO",
  "contact" => {
    "company"  => "TJRM",
    "address1" => "123 N. Main",
    "address2" => "",
    "city"     => "Tulsa",
    "state"    => "OK",
    "zip"      => "74137",
    "country"  => "US",
    "phone"    => ""
  },
  "campaign_defaults" => {
    "from_name"  => "Tim Morgan",
    "from_email" => "tim@timmorgan.org",
    "subject"    => "test",
    "language"   => "English"
  },
  # ...
  "stats" => {
    "member_count" => 0,
    # ...
  },
  "_links" => [
    {
      "rel"          => "self",
      "href"         => "https://us2.api.mailchimp.com/3.0/lists/abc123abc1",
      "method"       => "GET",
      "targetSchema" => "https://us2.api.mailchimp.com/schema/3.0/Lists/Instance.json"
    },
    # ...
  ]
}
```

## method_missing(endpoint_name)

Anything you append to the end gets handled by `method_missing` and returns a new Endpoint object.
This is what allows this gem work with any endpoint without the library code needing to be updated.

## \[](endpoint_name_or_id)

Similar to the `method_missing` above, `endpoint['whatever']` and `endpoint[123]` are handled dynamically,
which allows you to build endpoints containing numbers, hyphens, or other characters not allowed by a
Ruby method name.

## get(params = {})

`get()` works for a collection (index) and a single resource (show).

```ruby
# collection
api.lists['abc123'].members.get(count: 25)
# => { members: array_of_resources }

# single resource
api.lists['abc123'].members['cde345'].get
# => resource_hash
```

If the endpoint accepts parameters, you can pass them in the `params` hash.

## post(body = {})

`post()` sends a POST request to create a new resource.

```ruby
api.lists['abc123'].members.post(...)
# => resource_hash
```

The parameter given is sent as the body of the request. See [Example](#example) above.

## patch(body = {})

`patch()` sends a PATCH request to update an existing resource.

```ruby
api.lists['abc123'].members['cde345'].patch(...)
# => resource_hash
```

The parameter given is sent as the body of the request. See [Example](#example) above.

## delete()

`delete()` sends a DELETE request to delete an existing resource. This method returns `true` if the delete was successful.

```ruby
api.lists['abc123'].members['cde345'].delete
# => true
```

## Errors

The following errors may be raised by the library, depending on the API response status code.

| HTTP Status Codes   | Specific Error Class                      | Generic Error Class               |
| ------------------- | ----------------------------------------- | --------------------------------- |
| 400                 | `MailChimp3::Errors::BadRequest`          | `MailChimp3::Errors::ClientError` |
| 401                 | `MailChimp3::Errors::Unauthorized`        | `MailChimp3::Errors::ClientError` |
| 403                 | `MailChimp3::Errors::Forbidden`           | `MailChimp3::Errors::ClientError` |
| 404                 | `MailChimp3::Errors::NotFound`            | `MailChimp3::Errors::ClientError` |
| 405                 | `MailChimp3::Errors::MethodNotAllowed`    | `MailChimp3::Errors::ClientError` |
| 422                 | `MailChimp3::Errors::UnprocessableEntity` | `MailChimp3::Errors::ClientError` |
| other 4xx errors    | `MailChimp3::Errors::ClientError`         |                                   |
| 500                 | `MailChimp3::Errors::InternalServerError` | `MailChimp3::Errors::ServerError` |
| other 5xx errors    | `MailChimp3::Errors::ServerError`         |                                   |

The exception object has the following methods:

| Method  | Content                                  |
| ------- | ---------------------------------------- |
| status  | HTTP status code returned by the server  |
| message | the message returned by the API          |
| details | the full response returned by the server |

The `message` should be a simple string given by the API, e.g. "Resource Not Found".

`details` is a Ruby hash containing all the details given by the server, and looks like this:

```ruby
{
  "type"     => "http://kb.mailchimp.com/api/error-docs/400-invalid-resource",
  "title"    => "Invalid Resource",
  "status"   => 400,
  "detail"   => "The resource submitted could not be validated. For field-specific details, see the 'errors' array.",
  "instance" => "286179fe-f3dc-4c03-8c14-1021cf0191a2",
  "errors" => [
    {
      "field"   => "",
      "message" => "Required fields were not provided: permission_reminder, campaign_defaults"
    }
  ]
}
```

Alternatively, you may rescue `MailChimp3::Errors::BaseError` and branch your code based on
the status code returned by calling `error.status`.

## Copyright & License

Copyright 2021, Planning Center. Licensed MIT.


CHANGELOG.md
# 1.5.2 (April 4, 2023)

* Fix: Forward keyword args to `MailChimp3#new` in Ruby 3.x compatible way (#17)

# 1.5.1 (December 10, 2021)

* Fix: Resolve faraday deprecation warning
* Fix: Loosen `faraday` dependency to 1.x
* Fix: Loosen `webmock` dependency to 3.x
* Fix: Move Faraday adapter declaration to the end

# 1.5.0 (April 6, 2021)

* Feature: allow using a state parameter in the OAuth flow

# 1.4.0 (June 22, 2020)

* Fix: Don't try to parse JSON on POST request for a 204 response
* Chore: Bump minimum Excon version for CVE-2019-16779

# 1.3.3 (May 19, 2019)

* Fix: Loosen up Faraday dependency

# 1.3.2 (April 2, 2019)

* Fix: Loosen up Excon dependency

# 1.3.1 (August 8, 2017)

* Chore: Be more liberal with oauth2 gem version

# 1.3.0 (November 14, 2016)

* Chore: Bump oauth2 to 1.2

# 1.2.0 (October 28, 2015)

* Fix: raise `MailChimp3::Errors::ServerError` when response JSON cannot be parsed

# 1.1.0 (July 6, 2015)

* Feature: add method (`#v2`) for working with the MailChimp 2.0 API

# 1.0.0 (June 29, 2015)

Initial release.


spec/spec_helper.rb
require 'webmock/rspec'
require 'pry'

require_relative '../lib/mailchimp3'

RSpec.configure do |config|
  config.expect_with :rspec do |expectations|
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  config.mock_with :rspec do |mocks|
    mocks.verify_partial_doubles = true
  end

  config.color = true
  config.order = 'random'
  config.filter_run focus: true
  config.run_all_when_everything_filtered = true
end


spec/mailchimp3_spec.rb
require_relative 'spec_helper'

describe MailChimp3 do
  describe '#new' do
    it 'smoke test' do
      expect(subject.new(basic_auth_key: 'key-us2').url).not_to be_nil
    end
  end

  describe '#config' do
    it 'returns configuration object' do
      subject.config.client_id = 'foo'
      subject.config.client_secret = 'bar'
      expect(subject.config.client_id).to eq('foo')
      expect(subject.config.client_secret).to eq('bar')
    end
  end

  describe '#oauth' do
    it 'returns a MailChimp3::OAuth instance' do
      expect(subject.oauth).to be_a(MailChimp3::OAuth)
    end
  end
end


spec/mailchimp3/oauth_spec.rb
require_relative '../spec_helper'

MailChimp3.config.client_id = 'foo'
MailChimp3.config.client_secret = 'bar'

describe MailChimp3::OAuth do
  subject { MailChimp3.oauth }

  describe '#authorize_url' do
    it 'returns the authorization url' do
      url = subject.authorize_url(
        redirect_uri: 'http://example.com/oauth/callback'
      )
      expect(url).to eq('https://login.mailchimp.com/oauth2/authorize?client_id=foo&redirect_uri=http%3A%2F%2Fexample.com%2Foauth%2Fcallback&response_type=code')
    end

    it 'passes through an optional state param' do
      url = subject.authorize_url(
        redirect_uri: 'http://example.com/oauth/callback',
        state: "123456"
      )
      expect(url).to eq('https://login.mailchimp.com/oauth2/authorize?client_id=foo&redirect_uri=http%3A%2F%2Fexample.com%2Foauth%2Fcallback&response_type=code&state=123456')
    end
  end

  describe '#complete_auth' do
    before do
      stub_request(:post, 'https://login.mailchimp.com/oauth2/token')
        .with(body: {
          'client_id'     => 'foo',
          'client_secret' => 'bar',
          'code'          => '1234567890',
          'grant_type'    => 'authorization_code',
          'redirect_uri'  => 'http://example.com/oauth/callback'
        })
        .to_return(
          status: 200,
          body: {
            access_token: '925680f04933b28f128d721fdf8949fa',
            expires_in: 0,
            scope: nil
          }.to_json,
          headers: {
            'Content-Type' => 'application/json'
          }
        )
        stub_request(:get, 'https://login.mailchimp.com/oauth2/metadata')
          .with(headers: {
            'Authorization' => 'Bearer 925680f04933b28f128d721fdf8949fa'
          })
          .to_return(
            status: 200,
            body: {
              dc: 'us2',
              role: 'owner',
              accountname: 'timmorgan',
              user_id: 2472146,
              login: {
                email: 'tim@timmorgan.org',
                avatar: nil,
                login_id: 2472146,
                login_name: 'timmorgan',
                login_email:'tim@timmorgan.org'
              },
              login_url: 'https://login.mailchimp.com',
              api_endpoint: 'https://us2.api.mailchimp.com'
            }.to_json)
    end

    it 'stores the auth token and metadata' do
      url = subject.authorize_url(
        redirect_uri: 'http://example.com/oauth/callback'
      )
      hash = subject.complete_auth(
        '1234567890',
        redirect_uri: 'http://example.com/oauth/callback'
      )
      expect(hash[:token]).to be_an(OAuth2::AccessToken)
      expect(hash[:token_string]).to eq('925680f04933b28f128d721fdf8949fa')
      expect(hash[:metadata]).to be_a(Hash)
      expect(hash[:metadata][:dc]).to eq('us2')
    end
  end
end


lib/mailchimp3/oauth.rb
require 'oauth2'
require 'json'

module MailChimp3
  class OAuth
    def initialize
      @oauth = OAuth2::Client.new(
        MailChimp3.config.client_id,
        MailChimp3.config.client_secret,
        site: 'https://login.mailchimp.com',
        authorize_url: '/oauth2/authorize',
        token_url: '/oauth2/token'
      )
    end

    def authorize_url(redirect_uri:, state: nil)
      params = {
        redirect_uri: redirect_uri,
        state: state,
      }.compact

      @oauth.auth_code.authorize_url(params)
    end

    def complete_auth(code, redirect_uri:)
      token = @oauth.auth_code.get_token(
        code,
        redirect_uri: redirect_uri
      )
      {
        token: token,
        token_string: token.token,
        metadata: metadata(token)
      }
    end

    private

    def metadata(token)
      JSON.parse(token.get('/oauth2/metadata').body).tap do |hash|
        hash.keys.each do |key|
          hash[key.to_sym] = hash.delete(key)
        end
      end
    end
  end
end


lib/mailchimp3/errors.rb
module MailChimp3
  module Errors
    class AuthRequiredError < StandardError; end
    class DataCenterRequiredError < StandardError; end

    class BaseError < StandardError
      attr_reader :status, :details

      def initialize(body:, status:)
        @status = status
        @message = if body.is_a?(Hash) && body['error']
                     "#{body['name']}: #{body['error']}"
                   elsif body.is_a?(Hash)
                     "#{body['title']}: #{body['detail']}"
                   else
                     body.to_s
                   end
        @details = body if body.is_a?(Hash)
      end

      def to_s
        @message
      end

      def inspect
        "<#{self.class.name} status=#{@status} message=#{@message}>"
      end
    end

    class ClientError              < BaseError;   end # 400..499
    class BadRequest               < ClientError; end # 400
    class Unauthorized             < ClientError; end # 401
    class Forbidden                < ClientError; end # 403
    class NotFound                 < ClientError; end # 404
    class MethodNotAllowed         < ClientError; end # 405
    class UnprocessableEntity      < ClientError; end # 422

    class ServerError              < BaseError;   end # 500..599
    class InternalServerError      < ServerError; end # 500
  end
end


lib/mailchimp3/endpoint.rb
require 'faraday'
require 'json'

module MailChimp3
  class Endpoint
    attr_reader :url, :last_result

    def initialize(oauth_access_token: nil, basic_auth_key: nil, dc: nil, url: nil, version: 3)
      @oauth_access_token = oauth_access_token
      @basic_auth_key = basic_auth_key
      @dc = dc
      @dc ||= @basic_auth_key.split('-').last if @basic_auth_key
      @url = url || _build_url
      @version = version
      fail Errors::DataCenterRequiredError, 'You must pass dc.' unless @dc || @url
      @cache = {}
    end

    def method_missing(endpoint_name, *_args)
      _build_endpoint(endpoint_name.to_s)
    end

    def [](endpoint_name_or_id)
      _build_endpoint(endpoint_name_or_id.to_s)
    end

    def get(params = {})
      @last_result = _connection.get(@url, params)
      _build_response(@last_result)
    end

    def post(body = {})
      @last_result = _connection.post(@url) do |req|
        body[:apikey] = @oauth_access_token || @basic_auth_key if @version == 2
        req.body = body.to_json
      end
      if @last_result.status == 204
        true
      else
        _build_response(@last_result)
      end
    end

    def patch(body = {})
      @last_result = _connection.patch(@url) do |req|
        req.body = body.to_json
      end
      _build_response(@last_result)
    end

    def put(body = {})
      @last_result = _connection.put(@url) do |req|
        req.body = body.to_json
      end
      _build_response(@last_result)
    end

    def delete
      @last_result = _connection.delete(@url)
      if @last_result.status == 204
        true
      else
        _build_response(@last_result)
      end
    end

    def v2
      self.class.new(
        url: _build_v2_url,
        basic_auth_key: @basic_auth_key,
        oauth_access_token: @oauth_access_token,
        version: 2
      )
    end

    private

    def _build_response(result)
      body = _parse_body(result)
      case (status = result.status)
      when 200..299
        body
      when 400
        fail Errors::BadRequest.new(status: status, body: body)
      when 401
        fail Errors::Unauthorized.new(status: status, body: body)
      when 403
        fail Errors::Forbidden.new(status: status, body: body)
      when 404
        fail Errors::NotFound.new(status: status, body: body)
      when 405
        fail Errors::MethodNotAllowed.new(status: status, body: body)
      when 422
        fail Errors::UnprocessableEntity.new(status: status, body: body)
      when 400..499
        fail Errors::ClientError.new(status: status, body: body)
      when 500
        fail Errors::InternalServerError.new(status: status, body: body)
      when 500..599
        fail Errors::ServerError.new(status: status, body: body)
      else
        fail "unknown status #{status}"
      end
    end

    def _parse_body(result)
      JSON.parse(result.body)
    rescue JSON::ParserError
      raise Errors::ServerError.new(status: result.status, body: result.body)
    end

    def _build_endpoint(path)
      @cache[path] ||= begin
        self.class.new(
          url: File.join(url, path.to_s),
          basic_auth_key: @basic_auth_key,
          oauth_access_token: @oauth_access_token,
          version: @version
        )
      end
    end

    def _build_url
      "https://#{@dc}.api.mailchimp.com/3.0/"
    end

    def _build_v2_url
      "https://#{@dc}.api.mailchimp.com/2.0/"
    end

    def _connection
      @connection ||= Faraday.new(url: url) do |faraday|
        if @basic_auth_key
          faraday.request :basic_auth, '', @basic_auth_key
        elsif @oauth_access_token
          faraday.headers['Authorization'] = "Bearer #{@oauth_access_token}"
        else
          fail Errors::AuthRequiredError, "You must specify either HTTP basic auth credentials or an OAuth2 access token."
        end
        faraday.adapter :excon
      end
    end
  end
end
